telemetry/telemetrybuffer.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2017 Microsoft. All rights reserved.
+// MIT License
+
+package telemetry
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// FdName - file descriptor name
+// Delimiter - delimiter for socket reads/writes
+// HostNetAgentURL - host net agent url of type payload
+// DefaultDncReportsSize - default DNC report slice size
+// DefaultCniReportsSize - default CNI report slice size
+// DefaultNpmReportsSize - default NPM report slice size
+// DefaultInterval - default interval for sending payload to host
+const (
+	FdName          = "azure-telemetry"
+	Delimiter       = '\n'
+	HostNetAgentURL = "http://169.254.169.254/machine/plugins?comp=netagent&type=payload"
+	DefaultInterval = 1 * time.Minute
+)
+
+// TelemetryBuffer object
+type TelemetryBuffer struct {
+	client      net.Conn
+	listener    net.Listener
+	connections []net.Conn
+	payload     Payload
+	fdExists    bool
+	connected   bool
+	data        chan interface{}
+	cancel      chan bool
+}
+
+// Payload object holds the different types of reports
+type Payload struct {
+	DNCReports []DNCReport
+	CNIReports []CNIReport
+	NPMReports []NPMReport
+}
+
+// NewTelemetryBuffer - create a new TelemetryBuffer
+func NewTelemetryBuffer() (*TelemetryBuffer, error) {
+	var tb TelemetryBuffer
+	tb.data = make(chan interface{})
+	tb.cancel = make(chan bool, 1)
+	tb.connections = make([]net.Conn, 1)
+	err := tb.Listen(FdName)
+	if err != nil {
+		tb.fdExists = strings.Contains(err.Error(), "exist") || strings.Contains(err.Error(), "in use")
+	} else {
+		// Spawn server goroutine to handle incoming connections
+		go func() {
+			for {
+				// Spawn worker goroutines to communicate with client
+				conn, err := tb.listener.Accept()
+				if err == nil {
+					tb.connections = append(tb.connections, conn)
+					go func() {
+						for {
+							reportStr, err := read(conn)
+							if err == nil {
+								var tmp map[string]interface{}
```

Comment:
Can you change the name of this variable from tmp to something more meaningful?

Reply:
will change when we discover a better name